portal/views/reporting.py

```diff
@@ -152,3 +158,124 @@ def overdue(qstats):
+      401:
+        description:
+          if missing valid OAuth token or if the authorized user lacks
+          permission to view requested user_id
+
+    """
+
+    if request.args.get('as_of_date'):
+        as_of_date = FHIR_datetime.parse(request.args.get('as_of_date'))
+    else:
+        as_of_date = datetime.utcnow()
+
+    # Obtain list of qualifying patients (not marked test)
+    # TODO: refactor this common query need to model and replace
+    # the less efficient similar usage elsewhere...
+    test_user_ids = UserRoles.query.join(Role).filter(
+        UserRoles.role_id == Role.id).filter(
+        Role.name == ROLE.TEST.value).with_entities(UserRoles.user_id)
+    patients = User.query.filter(User.active.is_(True)).join(
+        UserRoles).filter(User.id == UserRoles.user_id).join(
+        Role).filter(Role.name == ROLE.PATIENT.value).filter(
+        ~User.id.in_(test_user_ids))
+
+    # If limited by org - grab org and all it's children, and refine query
+    org_id = request.args.get('org_id')
+    if org_id:
+        limit_orgs = OrgTree().here_and_below_id(organization_id=org_id)
+        patients = patients.join(UserOrganization).filter(
+            User.id == UserOrganization.user_id).filter(
+            UserOrganization.organization_id.in_(limit_orgs))
+
+    # Todo: confirm current_user has view on all patients
+    results = []
+    for patient in patients:
+        if not patient.organizations.first():
+            # Very unlikely we want to include patients w/o at least
+            # one org, skip this patient
+            continue
+
+        qb_stats = QB_Status(user=patient, as_of_date=as_of_date)
+        row = {
+            'user_id': patient.id,
+            'site': patient.organizations.first().name,
+            'status': str(qb_stats.overall_status)}
+
+        consent = latest_consent(user=patient)
+        if consent:
+            row['consent'] = FHIR_datetime.as_fhir(consent.acceptance_date)
+
+        study_id = patient.external_study_id
+        if study_id:
+            row['study_id'] = study_id
+
+        current = qb_stats.current_qbd()
+        previous = qb_stats.prev_qbd
+        if current:
+            row['visit'] = visit_name(current)
+        elif previous and qb_stats.overall_status == OverallStatus.expired:
+            # It's the previous visit that expired, not the current
+            row['visit'] = visit_name(previous)
+        results.append(row)
+
+    if request.args.get('format', 'json').lower() == 'csv':
+        def gen(items):
+            desired_order = ['study_id', 'status', 'visit', 'site', 'consent']
+            yield ','.join(desired_order) + '\n'  # header row
+            for i in items:
+                yield ','.join([i.get(k, "") for k in desired_order]) + '\n'
+
+        return Response(
+            gen(results),
+            mimetype='text/plain',
+            headers={'Content-Disposition': 'attachment;filename=status.csv'}
```

Comment:
just icing: would be nice if this filename had a timestamp (or date at least), and the org param if it exists